docs/contributing/smart-component.md

```diff
@@ -2,9 +2,199 @@
+
+## Building a Smart Component
+
+### 1. Selector and Context Parameters
+
+First, define your smart component with a selector and context parameters:
+
+```js
+defineSmartComponent({
+  selector: 'cc-my-component', // CSS selector used to find and bind to all instances of this component in the DOM
+  params: {
+    // Define expected context parameters
+    apiConfig: { type: Object },
+    userId: { type: String },
+    isAdmin: { type: Boolean, optional: true }
+  },
+  onContextUpdate: ...
+});
+```
+
+A smart component reacts to changes in its context by receiving updates through the `onContextUpdate` callback.
+The `params` section defines the specific pieces of context data that the component needs to function and interact with APIs.
+Parameters can be marked as optional with `optional: true` if they are not strictly required.
+**At least one parameter must be mandatory (non-optional) - you cannot make all parameters optional**.
+The `type` values inside `params` serve as documentation and have no functional impact on the component's behavior.
+
+### 2. State Updates and Data Fetching
+
+The `updateComponent` helper is a powerful tool for managing component state:
+
+```js
+// Pass a direct value to set the property
+updateComponent('state', { state: 'loading' });
+
+// Pass a function to modify the property with Immer
+updateComponent('state', (state) => {
+  state.saving = true;
+  state.lastUpdate = Date.now();
+});
+```
+
+Using [Immer](https://github.com/immerjs/immer) under the hood, `updateComponent` lets you write immutable updates in a mutable style.
+This allows for more concise and readable state updates while still triggering Lit's change detection.
+
+When fetching data on context updates, start by destructuring the context parameters:
+
+```js
+onContextUpdate({ component, context, updateComponent, signal }) {
+  const { apiConfig, userId } = context;
+
+  // Set loading state
+  updateComponent('state', { state: 'loading' });
+
+  // Fetch data using destructured context params
+  fetchData(apiConfig, userId, signal)
+    .then((data) => {
+      updateComponent('state', {
+        state: 'loaded',
+        value: data
+      });
+    })
+    .catch((error) => {
+      console.error(error);
+      updateComponent('state', { state: 'error' });
+    });
+}
+```
+
+The `signal` parameter allows proper cleanup when the context changes again. When a new context update occurs, the previous request will be aborted, preventing race conditions with stale data.
+
+Note:
+[Immer](https://github.com/immerjs/immer) freezes objects to ensure immutability.
+If you try to directly mutate an Immer-produced object elsewhere in your code, you'll get an error indicating you cannot modify readonly properties.
+This is error is legitimate since you should never directly mutate Lit reactive properties, even though plain JavaScript would technically allow it.
```

Comment:
```suggestion
This error is legitimate since you should never directly mutate Lit reactive properties, even though plain JavaScript would technically allow it.
```

Reply:
oupsie thanks!